### PR TITLE
refactor(j-s): Transition Side Effects

### PR DESCRIPTION
**Pull request by gudjong**
# Transition Side Effects

[Rannsaka villur í R-málum](https://app.asana.com/0/1199153462262248/1208647533881237/f)

## What

- Moves transition side effects from controllers to the case state machine.

## Why

- Cleaner code.

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

## Summary by CodeRabbit

- **New Features**
	- Streamlined case transition handling for improved clarity and maintainability.
	- Enhanced error handling for court record and ruling signature requests.
	- Simplified transition logic in controllers for better usability.

- **Bug Fixes**
	- Improved handling of invalid state transitions with structured error responses.

- **Tests**
	- Expanded test coverage for case transitions, including appeal scenarios and notification checks.
	- Enhanced user role context in tests for more accurate behavior validation.

### CODE REVIEW

File: apps/judicial-system/backend/src/app/modules/case/state/case.state.spec.ts [hunk 1]
```diff
@@ -17,7 +23,16 @@ describe('Transition Case', () => {
...
+        const act = () =>
+          transitionCase(
+            CaseTransition.OPEN,
+            { id: uuid(), state: fromState, type } as Case,
+            {
+              id: uuid(),
+              role: UserRole.PROSECUTOR,
+              institution: { type: InstitutionType.PROSECUTORS_OFFICE },
+            } as User,
+          )
```
[the remark this answers is not in the document]
Comment (gudjong) [reply apps/judicial-system/backend/src/app/modules/case/state/case.state.spec.ts:35]: I am happy with my tests being like this.

File: apps/judicial-system/backend/src/app/modules/case/state/case.state.spec.ts [hunk 2]
```diff
@@ -1,14 +1,20 @@
...
   CaseAppealState,
   CaseState,
   CaseTransition,
   indictmentCases,
+  InstitutionType,
   investigationCases,
   restrictionCases,
+  User,
+  UserRole,
 } from '@island.is/judicial-system/types'
```
[the remark this answers is not in the document]
Comment (gudjong) [reply apps/judicial-system/backend/src/app/modules/case/state/case.state.spec.ts:16]: I am happy with my tests.

File: apps/judicial-system/backend/src/app/modules/case/state/case.state.ts [hunk 3]
```diff
@@ -325,25 +526,30 @@
...
+  update: UpdateCase = {},
 ): CaseStates {
-  if (isIndictmentCase(caseType)) {
-    return transitionIndictmentCase(transition, currentState)
+  const actor = !isProsecutionUser(user)
+    ? !isDefenceUser(user)
+      ? 'Nutral'
+      : 'Defence'
+    : 'Prosecution'
+
```
[the remark this answers is not in the document]
Comment (gudjong) [reply apps/judicial-system/backend/src/app/modules/case/state/case.state.ts]: Good idea, done.

File: apps/judicial-system/backend/src/app/modules/case/state/case.state.ts [hunk 4]
```diff
@@ -1,40 +1,49 @@
...
-}
+import { nowFactory } from '../../../factories'
+import { UpdateCase } from '../case.service'
+import { Case } from '../models/case.model'
 
-interface RequestCaseStates {
-  state?: RequestCaseState
-  appealState?: CaseAppealState
-}
+type Actor = 'Prosecution' | 'Defence' | 'Nutral'
```
[the remark this answers is not in the document]
Comment (gudjong) [reply apps/judicial-system/backend/src/app/modules/case/state/case.state.ts]: Thanks, done.

File: apps/judicial-system/backend/src/app/modules/case/state/case.state.ts [hunk 5]
```diff
@@ -325,25 +526,33 @@ const transitionRequestCase = (
...
-  if (isRequestCase(caseType)) {
-    return transitionRequestCase(transition, currentState, currentAppealState)
+  if (isRequestCase(theCase.type)) {
+    return transitionRequestCase(transition, theCase, update, actor)
   }
 
   throw new ForbiddenException(
-    `The transition ${transition} cannot be applied to a ${caseType} case`,
+    `The transition ${transition} cannot be applied to a ${theCase.type} case`,
   )
```
[the remark this answers is not in the document]
Comment (gudjong) [reply apps/judicial-system/backend/src/app/modules/case/state/case.state.ts:557]: I had not noticed the type discrepancy. I have fixed this another way.

File: apps/judicial-system/backend/src/app/modules/case/state/case.state.ts [hunk 6]
```diff
@@ -105,35 +145,88 @@ const indictmentCaseStateMachine: Map<
...
+      return transitionRequestCase(
+        CaseTransition.APPEAL,
+        theCase,
+        newUpdate,
+        prosecutorAppealedInCourt ? 'Prosecution' : 'Defence',
+      )
+    }
+
+    return newUpdate
+  }
```
[the remark this answers is not in the document]
Comment (gudjong) [reply apps/judicial-system/backend/src/app/modules/case/state/case.state.ts:195]: I am happy with this as is.

**Review by oddsson — APPROVED** [review]

File: apps/judicial-system/backend/src/app/modules/case/state/case.state.ts [hunk 7]
```diff
@@ -325,25 +526,33 @@ const transitionRequestCase = (
     )
   }
 
-  // Since the state machine is a global constant, we spread the result before returning it to avoid accidental mutations
-  return { ...rule.to } as CaseStates
+  return rule.transition(update, theCase, actor) as UpdateCase
```
[the remark this answers is not in the document]
Comment (gudjong) [reply apps/judicial-system/backend/src/app/modules/case/state/case.state.ts]: I have fixed this another way.

**Pull request merged**